# 4 Pandas Introduction

This notebook will introduce you to the `pandas` library which facilitates the use of DataFrames, a data object which behaves a lot like a table.
<br>
<br>You will be instructed on how to construct and manipulate these DataFrames to facilitate information storage and retrieval.

---

In notebook 1, you were reminded how a single dataset can be passed to `matplotlib.pyplot.plot()` as two lists, one containing x values and one containing y values. An example that will be used a lot later in this exercise is the IR spectrum, with the x variable being wavenumber and y being % Transmittance.
<br>
<br>While, in principle, it is possible to store many datasets in the same way, each with a corresponding x and y list (differently named for each dataset), this method of data storage very quickly becomes unwieldy. In cases where many datasets share a list of x values, it would make much more sense to store different y lists within a single data object and retrieve entries from that object as needed. The `pandas` library allows for the creation an object called a DataFrame to aid in this.
<br>
<br>At its core, a DataFrame is a multi-dimensional matrix, presented in an easy to read table. A DataFrame consists of a series of columns, each with a title/name, and a series of rows, each with a label/index. The row labels are called an index, so each value in each column across a row has the same index value.
<br>
<br>The use of an index works well for storing multiple sets of IR data, since, provided all data were recorded in the same manner, each wavenumber recorded will have the same index.

✏️ Import the `pandas` library, so that you can make your own DataFrame object.

*Hint: You might want to make use of the `as` keyword, so that you don't have to write 'pandas' out so much. A common abbreviation for `pandas` is `pd`.*

In [4]:
import pandas as pd


A DataFrame is a data structure, like a variable or a list. Instead of holding a single value (like a variable), or a series of values (like a list), it holds a table of values.
<br>
<br>A `pandas` DataFrame is defined in the following manner:
<br>
<br>`df = pd.DataFrame(data=[], index =[], columns=[])`
<br>
<br>`data` takes the entries that will be held in the body of the table. If the table is to have more than one column, the data should be provided as lists corrsponding to each row, containing the entries running from left to right.
<br>
<br>`index` is a list of row titles. These can be numbers for reference, or strings, for example.
<br>
<br>`columns` is a list of column titles. Again, these can be numbers or strings, for example.
<br>
<br>A very basic 2x2 example is provided by the code below.

In [2]:
df = pd.DataFrame(data=[["Entry 1 1","Entry 1 2"],["Entry 2 1","Entry 2 2"]],index=["Row 1","Row 2"],columns=["Column 1","Column 2"])
print(df)

        Column 1   Column 2
Row 1  Entry 1 1  Entry 1 2
Row 2  Entry 2 1  Entry 2 2


✏️ Make a new 'DataFrame' object, with a sensible name. Use the documentation to show you which arguments to use. Try to recreate the melting/boiling point table below.
>|       | Ethanol      | Ethane |
>| -----------      | ----------- | ----------- |
>| **Melting point**      | -114      | -183       |
>| **Boiling point**      | 78   | -89        |

In [13]:
df=pd.DataFrame(data=[[-114,-183],[78,-89]],index=['Melting point','Boiling point'],columns=['Ethanol','Ethane'])
print(df)

               Ethanol  Ethane
Melting point     -114    -183
Boiling point       78     -89


Some terms relating to machine learning and DataFrames that will come up later, but are worth mentioning from the outset.
>
>**Samples -** These are overall sets of data, i.e. columns in the DataFrame that you reproduced above.
>
>**Features -** These are subcategories which the data in each sample are divided into and correspond to the rows of the DataFrame above (**Melting point** and **Boiling point**). For IR spectra, these would be the wavenumber at which each transmittance is measured.
>
>**Values -** These are the entries in the table itself, i.e. the temperatures recorded in the DataFrame above. For IR spectra, these would correspond to the measured % Transmittance values.

✏️ DataFrame objects have a useful method called `.head()`, which prints out the first 5 lines of the DataFrame. Use the `head` method to see if your DataFrame is correct. Make adjustments to your original definition, if it is not.

In [10]:
df.head()

Ethanol  Ethane
Melting point     -114    -183
Boiling point       78     -89

We can extract a column from a DataFrame in a similar way to that in which we extract a single value from a list.
<br>
<br>For a list `constants = [3.14, 8.314, 300000000]`, `constants[0]` gives `3.14`, the 0th element of the list.
<br>
<br>For a DataFrame, `dataframe["column_name"]` will pull out the column with that name.

✏️ What are the melting and boiling points of *ethane*? Extract the relevant column from your previously defined DataFrame.

In [17]:
df['Ethane']

Melting point   -183
Boiling point    -89
Name: Ethane, dtype: int64

**Note:**
>The single column a DataFrame returned from this command is also called a **Series**.

The `.loc[index_value]` method pulls a *row* (with index "index_value") from a DataFrame.

✏️ What are the melting points of ethanol and ethane? Use the `.loc[]` method to extract these values.

In [18]:
df.loc['Melting point']

Ethanol   -114
Ethane    -183
Name: Melting point, dtype: int64

✏️ Try putting these two things together to pull out just the melting point of ethanol.

In [23]:
df['Ethanol'].loc['Melting point']

-114

The `.iloc[i]` method is very similar to `.loc[index_value]`, except it gives you the `i-1`th row, rather than the row with index `index_value`. This can be useful, for example, if you remember the row number that contains a data entry, but not the name that you have given to the row.

✏️ Given that it is contained within the **first** row of the DataFrame, use the `iloc[]` method to pull out just the melting point of ethanol.

In [27]:
df['Ethanol'].iloc[0]

-114

---

We're now going to look at some actual IR data.
<br>
<br>We've seen that we can create a DataFrame from a list of lists (a bit like a matrix):

In [28]:
df = pd.DataFrame([[1, 2], [3, 4]], columns=["Col1", "Col2"])
df.head()

Col1  Col2
0     1     2
1     3     4

But when you have thousands of datapoints (like in an IR spectrum), typing this out this gets unmanageable.
<br>
<br> Instead, we can use the `read_csv()` function of the `pandas` library to take in the data for your DataFrame from a file. This could be a *.csv* file (a stripped down version of an Excel file), or a *.txt* file. The IR machines in the CTL produce a *.txt* file when you export, so we will proceed with this format.
<br>
<br>**Note:** The IR spectra provided as *.txt* files are named systematically such that the initial character of each file represents the subsitution pattern (ortho/meta/para), the name of the chemical comes second, and the repeat number (between 1 and 5) comes third, e.g. *m-nitrophenol_1.txt*.

✏️ You should have downloaded a folder called `IR Data` from Canvas. Take the the example file, `m-anisaldehyde_1.txt`. This is the first IR spectrum repeat (of five) of the the chemical, m-anisaldehyde. The *path* to this file will be 'the folder name/the file name'.
<br>
<br>Use the `read_csv()` function to create a new DataFrame from `m-anisaldehyde_1.txt`.
Don't worry about any arguments other than `path` for now.
<br>
<br>*Call your DataFrame `m_anisaldehyde_1` - note that there is no dash, as Python doesn'y accept dashes in variable names.*
<br>
<br>Use the `.head()` method to check if the DataFrame looks as expected.

In [37]:
m_anisaldehyde_1=pd.read_csv('data/m-anisaldehyde_1.txt')
m_anisaldehyde_1.head()

##TITLE=N1_m-anisaldehyde_1
0  ##DATA TYPE=INFRARED SPECTRUM
1                  ##XUNITS=1/CM
2                    ##YUNITS=%T
3       399.826377  \t92.012424 
4       400.183684  \t94.958885

The data's all there, but it looks a bit off. You can try opening the *.txt* file to see how the head of the DataFrame relates to the information stored in the file.
<br>
<br>You should notice that the first four rows of the *.txt* file do not contain spectral data. We can ignore these with the argument `skipsrows=4`.
<br>
<br>In lines 3 and 4 of your head above you will see that the wavenumber and the % Transmission have been grouped into the same cell entry. This is because the `read_csv()` function is expecting a CSV in which a comma would separate separate columns in a table. We can correct this by informing `read_csv` that the entries in our data are separated by spaces using the argument, `delimiter="\s+"`.
<br>
<br>In the same manner as above, we should also state some column headings for the DataFrame. When importing files, we use the argument `names=[]`, rather than `columns=[]` outlined earlier. The columns should be called "Wavelength/cm-1" and "% Transmittance". The first wavelengths should be around 400 cm-1, and the first transmittances around 90-100%.

✏️ By incorporating the arguments discussed above, read in the file again to give a DataFrame that looks like:
>|       | Wavelength /cm-1      | % Transmittance |
>| ----------- | ----------- | ----------- |
>| 0      | 399.826377      | 92.012424       |
>| 1      | 400.183684   | 94.958885        |

In [38]:
m_anisaldehyde_1=pd.read_csv('data/m-anisaldehyde_1.txt',skiprows=4,delimiter="\s+",names=["Wavelength/cm-1","% Transmittance"])
m_anisaldehyde_1.head()

Wavelength/cm-1  % Transmittance
0       399.826377        92.012424
1       400.183684        94.958885
2       400.540991        97.443201
3       400.898298        97.642822
4       401.255605        95.774480

✏️ How many rows does the `m_anisaldehyde_1` DataFrame have? Use the documentation linked to below to find an attribute that will return the dimensions of the DataFrame.
<br>
<br>https://pandas.pydata.org/docs/user_guide/basics.html#attributes-and-underlying-data
<br>
<br>*Hint: You'll get back something a bit like a list. Use the `[]` notation we've seen before to get a single number - the number of rows.*

In [41]:
m_anisaldehyde_1.shape[0]

10077

---

The DataFrame currently has an index column (0, 1, 2, ...), a wavelength column, and a trasmittance column. The index column is a bit redundant, and is not of any use. It would be more useful to have the wavelength column as the index.
<br>

✏️Use the documentation linked to below to find a method that will render the wavelength column as the index of the DataFrame.
<br>
<br>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html

In [68]:
df2=pd.DataFrame(data=m_anisaldehyde_1['% Transmittance'],index=m_anisaldehyde_1['Wavelength/cm-1'])
m_anisaldehyde_1=df2

✏️ What is the transmittance at 414.118652 $cm^{-1}$?

In [69]:
m_anisaldehyde_1.loc[414.118652]

% Transmittance    90.813631
Name: 414.118652, dtype: float64

---

✏️ Let's remind ourselves what the modified DataFrame looks like. Print out the first 5 lines.

In [70]:
m_anisaldehyde_1.head()

% Transmittance
Wavelength/cm-1                 
399.826377             92.012424
400.183684             94.958885
400.540991             97.443201
400.898298             97.642822
401.255605             95.774480

✏️ Imagine we had found that we'd accidentally used twice as much substance for the `m_anisaldehyde_1` spectrum as for the other repeats, so we wanted to halve all of the transmittance values.
<br>
<br>The naive way to do this would be to use a for loop to loop through each row, and manually change the value with `iloc`. It turns out that this is a very inefficient method.
<br>
<br>See what happens if you pull a column from a DataFrame and multiply it by 10.

In [71]:
m_anisaldehyde_1['% Transmittance']*10

Wavelength/cm-1
399.826377     920.12424
400.183684     949.58885
400.540991     974.43201
400.898298     976.42822
401.255605     957.74480
401.612912     968.25625
401.970219     980.04833
402.327525     965.83739
402.684832     946.50614
403.042139     928.04176
403.399446     937.58873
403.756753     935.63720
404.114060     921.98788
404.471367     942.49125
404.828674     936.53572
405.185980     899.31653
405.543287     884.69126
405.900594     901.31367
406.257901     920.32358
406.615208     904.41368
406.972515     891.75432
407.329822     908.74112
407.687128     929.46987
408.044435     946.90700
408.401742     949.18385
408.759049     933.13856
409.116356     929.14019
409.473663     937.29385
409.830970     926.54712
410.188276     912.74790
                 ...    
3989.688410    997.70712
3990.045717    997.46517
3990.403024    995.70907
3990.760330    996.84396
3991.117637    998.03842
3991.474944    997.17008
3991.832251    996.43128
3992.189558    995.31627
3992.5468

✏️ Hence, halve each transmittance value in `m_anisaldehyde_1`.

In [72]:
m_anisaldehyde_1['% Transmittance']*0.5

Wavelength/cm-1
399.826377     46.006212
400.183684     47.479442
400.540991     48.721601
400.898298     48.821411
401.255605     47.887240
401.612912     48.412813
401.970219     49.002416
402.327525     48.291870
402.684832     47.325307
403.042139     46.402088
403.399446     46.879436
403.756753     46.781860
404.114060     46.099394
404.471367     47.124563
404.828674     46.826786
405.185980     44.965826
405.543287     44.234563
405.900594     45.065683
406.257901     46.016179
406.615208     45.220684
406.972515     44.587716
407.329822     45.437056
407.687128     46.473493
408.044435     47.345350
408.401742     47.459193
408.759049     46.656928
409.116356     46.457009
409.473663     46.864692
409.830970     46.327356
410.188276     45.637395
                 ...    
3989.688410    49.885356
3990.045717    49.873258
3990.403024    49.785453
3990.760330    49.842198
3991.117637    49.901921
3991.474944    49.858504
3991.832251    49.821564
3992.189558    49.765814
3992.5468

✏️ Use the `.head()` method to look at the `m_anisaldehyde_1` DataFrame again - have your changes been applied?

In [73]:
m_anisaldehyde_1.head()

% Transmittance
Wavelength/cm-1                 
399.826377             92.012424
400.183684             94.958885
400.540991             97.443201
400.898298             97.642822
401.255605             95.774480

You should observe that the original DataFrame is unchanged. This is because the operations that you have been performing are not *in-place* operations, i.e. the original DataFrame was never overwritten.

✏️ Modify your code so that the `m_anisaldehyde_1` transmittance column is **permanently** halved.

*Hint: We can replace a whole column of a DataFrame using `df["column name"]` notation. Setting this equal to a new column/series will overwrite this column.*

In [77]:
m_anisaldehyde_1['% Transmittance']=m_anisaldehyde_1['% Transmittance']*0.5

✏️ Print the first 5 lines of `m_anisaldehyde_1`. It should look like:
>| Wavelength /cm-1      | % Transmittance |
>| ----------- | ----------- |
>| 399.826377      | 46.006212       |
>| 400.183684   | 47.479442        |

In [79]:
m_anisaldehyde_1.head()

% Transmittance
Wavelength/cm-1                 
399.826377             46.006212
400.183684             47.479442
400.540991             48.721601
400.898298             48.821411
401.255605             47.887240

---

Finally, it is often useful to be able to find the titles of the columns in our DataFrame (once we start importing hundreds of DataFrames, we'll need to give their transmittance columns all different titles, and it will be difficult to keep track of them otherwise).
<br>
<br>The `columns` attribute gives a list of the column titles of a DataFrame.

✏️ What does the `columns` attribute show for `m_anisaldehyde_1`?

In [80]:
m_anisaldehyde_1.columns

Index(['% Transmittance'], dtype='object')

Although it might appear with some extra decoration, attribute is basically just a list. Thus we can pull the column heading using an appropriate index of `.columns`.

✏️ By using indexing, pull the *transmittances* column from the m_anisaldehyde_1 DataFrame without explicitly writing out the column title, "% Transmittance".

In [83]:
m_anisaldehyde_1[m_anisaldehyde_1.columns[0]]

Wavelength/cm-1
399.826377     46.006212
400.183684     47.479442
400.540991     48.721601
400.898298     48.821411
401.255605     47.887240
401.612912     48.412813
401.970219     49.002416
402.327525     48.291870
402.684832     47.325307
403.042139     46.402088
403.399446     46.879436
403.756753     46.781860
404.114060     46.099394
404.471367     47.124563
404.828674     46.826786
405.185980     44.965826
405.543287     44.234563
405.900594     45.065683
406.257901     46.016179
406.615208     45.220684
406.972515     44.587716
407.329822     45.437056
407.687128     46.473493
408.044435     47.345350
408.401742     47.459193
408.759049     46.656928
409.116356     46.457009
409.473663     46.864692
409.830970     46.327356
410.188276     45.637395
                 ...    
3989.688410    49.885356
3990.045717    49.873258
3990.403024    49.785453
3990.760330    49.842198
3991.117637    49.901921
3991.474944    49.858504
3991.832251    49.821564
3992.189558    49.765814
3992.5468

---